phabricatoremails/mail.py

```diff
@@ -0,0 +1,214 @@
+                      unique_number: int, event):
+        template_name = 'secure/' + self._template_name
+        html_email, text_email = self._render(
+            css_inline, template_store, template_name, actor_name, revision, unique_number, event, **self._kwargs
+        )
+        subject = 'D{}: (secure bug {})'.format(revision.id, revision.bug.id)
+        return OutgoingEmail(self._template_name, subject, self._recipient_email, html_email, text_email)
+
+
+@dataclass
+class OutgoingEmail:
+    template_name: str
+    subject: str
+    to: str
+    html_contents: str
+    text_contents: str
+
+    def to_mime_message(self, from_address, include_target_in_subject=False):
+        msg = MIMEMultipart('alternative')
+        msg['From'] = from_address
+        msg['To'] = self.to
+        msg['Subject'] = '|{}| {}'.format(self.to, self.subject) if include_target_in_subject else self.subject
+
+        msg.attach(MIMEText(self.text_contents + '\n', 'plain'))
+        msg.attach(MIMEText(self.html_contents + '\n', 'html'))
+        return msg
+
+
+class FsMail:
+    _logger: Logger
+    _from_address: str
+    _index: int
+    _eml_path: pathlib.Path
+    _html_path: pathlib.Path
+    _text_path: pathlib.Path
+
+    def __init__(self, from_address: str, logger: Logger):
+        self._logger = logger
+        self._from_address = from_address
+        self._index = 0
+        self._eml_path = pathlib.Path('output/eml')
+        self._html_path = pathlib.Path('output/html')
+        self._text_path = pathlib.Path('output/text')
+        self._eml_path.mkdir(parents=True, exist_ok=True)
+        self._html_path.mkdir(parents=True, exist_ok=True)
+        self._text_path.mkdir(parents=True, exist_ok=True)
+
+    def send(self, emails: List[OutgoingEmail]):
+        for email in emails:
+            basefilename = '{}-to-{}'.format(self._index, email.to)
+            with open(str(self._eml_path.joinpath(basefilename + '.eml')), 'w') as file:
+                file.write(email.to_mime_message(self._from_address).as_string())
+            with open(str(self._html_path.joinpath(basefilename + '.html')), 'w') as file:
+                file.write(email.html_contents)
+            with open(str(self._text_path.joinpath(basefilename + '.text')), 'w') as file:
+                file.write(email.text_contents)
+
+            self._index += 1
+        self._logger.info('Recorded {} emails to the "output" directory'.format(len(emails)))
+
+
+class MailQueue:
+    def __init__(self, emails: List[OutgoingEmail]):
+        self._user_queues = {}
+        for email in emails:
+            user_queue = self._user_queues.setdefault(email.to, [])
+            user_queue.append(email)
+
+    async def send_queue(self, cb: Callable[[OutgoingEmail], None], to: str):
+        user_queue = self._user_queues[to]
+        for i, email in enumerate(user_queue):
+            if i != 0:
+                # We want to sleep between each email so that they arrive in order.
```

Comment:
This is a good point that I hummed on: these emails will be send one-after-the-other within seconds of each other. If we don't `sleep`, then they may arrive out-of-order, since the `Date` header doesn't go to the granularity of milliseconds.

Reply:
I would second the call to not add a `sleep` here --- it will keep the code cleaner, and realistically when a large number of emails is being sent, this sleep will not necessarily have the intended impact.

Picture a future scenario where emails will go into a buffering queue before being sent to the SMTP server, or a scenario in which the SMTP server buffers emails anyway due to maintenance or any other reason.

Reply:
When I tested locally, not having this sleep caused emails to arrive out-of-order.
I think that the best implementation here is:

1. Include the `Date` header
2. Sleep for one second between each email, so that the `Date` header has time to change

What do y'all think?

Reply:
Yes, add the `Date` header to match the timestamp the change was made in Phabricator.
No, we shouldn't sleep.